src/lib/Imp.hs

```diff
@@ -248,9 +248,14 @@ toImpOp (maybeDest, op) = case op of
     destToAtom dest
   PrimEffect refDest m -> do
     case m of
-      MAsk    -> returnVal =<< destToAtom refDest
-      MTell x -> addToAtom refDest x >> returnVal UnitVal
-      MPut x  -> copyAtom  refDest x >> returnVal UnitVal
+      MAsk      -> returnVal =<< destToAtom refDest
+      MExtend ~(Lam f) -> do
+        -- TODO: Update in-place?
+        refValue <- destToAtom refDest
+        result <- translateBlock mempty (Nothing, snd $ applyAbs f refValue)
+        copyAtom refDest result
+        returnVal UnitVal
```

Comment:
One notable regression is that previously doing `ref += val` with `val` having a table type would update `ref` in-place, while now we allocate a temporary which later gets copied back to `ref`s storage. I'm hoping that making it possible to write the monoidal combiners as stateful functions will help us optimize that significantly.